### PR TITLE
quick and dirty

### PR DESCRIPTION
**Pull request by autumn-ragland-quad**
image

### CODE REVIEW

File: GeneralJS/top-product.ts [hunk 1]
```diff
@@ -89,6 +89,31 @@ const orders = [
 ];
 
 function topProduct(state: string) {
+  // match state to state code
+  const stateCode = states.find(stateObject => stateObject.state == state)?.stateCode;
+
+  // all products for passed state
+  const stateProducts = orders.filter(order => order.stateCode == stateCode);
+
+  // pull each product
```
Comment (autumn-ragland-quad) [inline GeneralJS/top-product.ts]: I feel like I should loop through products and determine most repeats at the same time...we'll see if I push up another commit before our meeting 😬

File: GeneralJS/top-product.ts [hunk 2]
```diff
@@ -89,6 +89,31 @@ const orders = [
...
+  // this will get gross and annoying the MINUTE more products are added...
```
Comment (aflinky) [inline GeneralJS/top-product.ts]: 🙃 yes

File: GeneralJS/top-product.ts [hunk 3]
```diff
@@ -89,6 +89,43 @@ const orders = [
...
+    Sorted Products
+    {
+      '1': 2, 
+      '2': 1 
+    }
+  */
+
+  // find the product id with the highest number of products
+  // full transparency I cobbled this together from a couple of stack overflow comments also I'm still iterating twice (embarrassing)
+  const topProductId = Number(Object.keys(sortedProducts).reduce((a, b) => sortedProducts[a] > sortedProducts[b] ? a : b));
```
Comment (aflinky) [inline GeneralJS/top-product.ts]: This use of reduce runs through all the sortedProducts again (you can tell me what the time complexity is 😃)

File: GeneralJS/top-product.ts [hunk 4]
```diff
@@ -89,6 +89,43 @@ const orders = [
...
+  
+  // match product id to product details
+  const topProductDetails = products.find(product => product.productId == topProductId);
```
Comment (aflinky) [inline GeneralJS/top-product.ts:129]: You can also tell me the time complexity of this find

**Review by aflinky — COMMENTED** [review]
This solution technically works but it can still be optimized!

The comparison you make in your reduce can be used within your first for loop so you can avoid the next two loops (the reduce and the find)

File: GeneralJS/top-product.ts (hunk 4 above)
Comment (autumn-ragland-quad) [reply GeneralJS/top-product.ts:129]: O(n). As complex as the array is long but if the matching element is early in the array is can be faster than if it's at the end (assuming the array is a substantial length)

**Review by aflinky — APPROVED** [review]

**Pull request merged**